yt/frontends/cf_radial/data_structures.py

```diff
@@ -0,0 +1,269 @@
+from .fields import CFRadialFieldInfo
+
+
+class CFRadialGrid(AMRGridPatch):
+    _id_offset = 0
+
+    def __init__(self, id, index, level, dimensions):
+        super().__init__(id, filename=index.index_filename, index=index)
+        self.Parent = None
+        self.Children = []
+        self.Level = level
+        self.ActiveDimensions = dimensions
+
+    def __repr__(self):
+        return "CFRadialGrid_%04i (%s)" % (self.id, self.ActiveDimensions)
+
+
+class CFRadialHierarchy(GridIndex):
+    grid = CFRadialGrid
+
+    def __init__(self, ds, dataset_type="cf_radial"):
+        self.dataset_type = dataset_type
+        self.dataset = weakref.proxy(ds)
+        # for now, the index file is the dataset!
+        self.index_filename = self.dataset.parameter_filename
+        self.directory = os.path.dirname(self.index_filename)
+        # float type for the simulation edges and must be float64 now
+        self.float_type = np.float64
+        super().__init__(ds, dataset_type)
+
+    def _initialize_state_variables(self):
+        super()._initialize_state_variables()
+        self.num_grids = 1
+
+    def _detect_output_fields(self):
+        # This needs to set a self.field_list that contains all the available,
+        # on-disk fields. No derived fields should be defined here.
+        # NOTE: Each should be a tuple, where the first element is the on-disk
+        # fluid type or particle type.  Convention suggests that the on-disk
+        # fluid type is usually the dataset_type and the on-disk particle type
+        # (for a single population of particles) is "io".
+        self.field_list = []
+
+        # pull out some info for dynamically setting fields and units
+        fic = self.ds._field_info_class
+        known_fieldnames = [fld[0] for fld in fic.known_other_fields]
+
+        flds_to_add = tuple()  # entries to add to known_other_fields
+        warn_units = 0  # flag if we encounter units we cannot handle
+        for key in self.ds._handle.variables.keys():
+            if (
+                all(x in self.ds._handle[key].dims for x in ["time", "z", "y", "x"])
+                is True
+            ):
+                field_tup = ("cf_radial", key)
+                warn_units += key in fic.dBz_fields
+                self.field_list.append(field_tup)
+                if key not in known_fieldnames:
+                    # this field is not in known fields, add it now after some
+                    # sanitization of the units
+                    units = self.ds._handle[key].units
+
+                    if units in fic.units_not_handled:
+                        units = ""
+                        warn_units += 1
+
+                    for findstr, repstr in fic.unit_subs:
+                        units = units.replace(findstr, repstr)
+
+                    flds_to_add += ((key, (units, [], self.ds._handle[key].long_name)),)
+
+        if flds_to_add:
+            self.ds._field_info_class.known_other_fields += flds_to_add
```

Comment:
```suggestion
        self.ds._field_info_class.known_other_fields += tuple(flds_to_add)
```
Actually this feels very much like a hack and I'd like to understand your motivation to update an external class here. I'm convinced this isn't necessary or that at least it shouldn't be.

Reply:
ya, it was a bit hacky (this change was mine) -- the basic issue was just that netcdf's can have a lot of fields and writing out all possible fields in `known_other_fields` isn't really possible. Without that modification, fields added to `self.field_list` that don't exist in `known_other_fields` load OK but don't have any units attached. This was my attempt to dynamically pull out the netcdf attribute data in a way that the units and descriptions would be used by yt -- maybe something exists already to do this more cleanly?

Reply:
Thought a bit more about this and looked through `FieldInfoContainer` again. Maybe what I'm doing here actually belongs in `CFRadialFieldInfo.setup_fluid_fields()` with calls to `CFRadialFieldInfo.add_output_field()`.

Reply:
I remember doing something hacky with dust-related fields in the AMRVAC frontend (since there can be an arbitrary number of them). I think that the field recognition also isn't flexible enough for the real world but obviously changing it is way beyond the scope of this PR. In any case I could be inclined to approve of a hack if only it happened closer to where the problem lies (the FieldInfo subclass). Do you think it's possible ?

Reply:
yes! just worked out how to do this in `setup_fluid_fields` in a cleaner way similar to other frontends... Will push it up soon!

Reply:
Ok, this might be resolved now. it's at least a lot cleaner and follows what other frontends do.